### PR TITLE
Create a PAPI client able to import ICS

### PR DESCRIPTION
**Pull request by thomas-hilaire**

### CODE REVIEW

File: src/papi-client.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        private credentials: PapiCredentials
+    ) {}
+
+    public startBatch(): Promise<Response> {
+        if (this.currentBatchId) {
+            throw new Error("The following batch is already started: " + this.currentBatchId);
+        }
+
+        return Promise.fromCallback(callback => {
+                this.auth(post(this.papiUrl("/batches/"))).end(callback);
```
Comment (mbaechler) [inline src/papi-client.ts]: it looks like you repeat `this.auth(<verb>(this.papiUrl`everytime, don't you ?

File: src/papi-client.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const DEFAUT_DELAY_MS = 1000;
+
+export class PapiClient {
+
+    public currentBatchId: BatchId;
+
+    constructor(
+        private apiUrl: string,
```
Comment (thomas-hilaire) [inline src/papi-client.ts]: It lacks of some common types like UUID, URL, ..

File: src/papi-client.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        private domainUUID: string,
```
Comment (fvn-linagora) [inline src/papi-client.ts]: domainUUID has a different type than models.ts/UUID ?

File: src/papi-client.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                .send(event.MimeContent)
+                .end(callback);
+        });
+    }
+
+    public importAllICS(events: EventMessage[]): Promise<Response[]> {
+        return Promise.all(events.map(e => this.importICS(e)));
+    }
+
+    private auth(request) {
```
Comment (fvn-linagora) [inline src/papi-client.ts]: maybe the whole auth + url building could be extracted into its own module/class, exposing each http verb ?

File: tests/papi-client.spec.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+let mock = require("superagent-mocker")(superagent);
+
+function expectBatchStart() {
+    mock.post("http://obm.org/my-domain/batches/", () => {
+        return {body: {id: 123}};
+    });
+}
+
+function generateEvent(id?: number) {
```
Comment (fvn-linagora) [inline tests/papi-client.spec.ts]: function signature could make use of available types ?

File: tests/papi-client.spec.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                .then(papiClient.waitForBatchSuccess.bind(papiClient, 10))
+                .then(
+                    () => {
+                        expect(queryCount).to.equal(3);
+                        done();
+                    },
+                    done
+                );
+        });
+
```
Comment (fvn-linagora) [inline tests/papi-client.spec.ts]: do you handle the case where the batch response status is marked SUCCESS, but there are inconsistencies between operationCount and operationDone (for instance!) ? Servers can sometimes have awkward behaviors/response :)

File: src/papi-client.ts (hunk 2 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: done, everything is a string at the end but I don't want to validate given args ..

File: src/papi-client.ts (hunk 1 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: I already tried to make that better but without success. Now I've a better understanding of TS so I finally found a solution. Done.

File: src/papi-client.ts (hunk 3 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: The same, string type :)
I made a types.ts file to make a placeholder of all types used in the project. Done.

File: src/papi-client.ts (hunk 4 above)
Comment (thomas-hilaire) [inline src/papi-client.ts]: I made new shortcut methods `request` and `requestInBatch` in this class.

File: tests/papi-client.spec.ts (hunk 5 above)
Comment (thomas-hilaire) [inline tests/papi-client.spec.ts]: done

File: tests/papi-client.spec.ts (hunk 6 above)
Comment (thomas-hilaire) [inline tests/papi-client.spec.ts]: All batch operations are reported in a `operations` field. In later commits I verify the status of each operation to log and requeue message in error.

File: src/papi-client.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        private domainUUID: UUID,
+        private credentials: PapiCredentials
+    ) {}
+
+    public startBatch(): Promise<Response> {
+        if (this.currentBatchId) {
+            throw new Error("The following batch is already started: " + this.currentBatchId);
+        }
+
+        return Promise.fromCallback(callback => this.request(post, "/batches/").end(callback))
```
Comment (fvn-linagora) [inline src/papi-client.ts]: Could the explicit promise creation (from callbacks) be hidden in request(Batch) functions with typed returned value ?

Comment (thomas-hilaire) [inline src/papi-client.ts]: I did a `promisify` method taking a "request" and returning a "promise".
I need to keep the `request` method returning a SuperAgentRequest to be able to customize it .. see the `importICS` or the `waitForBatchSuccess` methods.

**Comment by fvn-linagora** [comment]
:+1:

**Pull request merged**